### ISSUE TITLE
RFC: Refinements

### ISSUE THREAD

**Issue by mmiszy** [typeofweb/schema#45]
# RFC: Refinements

<details open>
<summary>Version 2, 2020-01-27</summary>

## Goals and use cases

1. Introduce a mechanism which will allow users to extend schema with their own rules of validation without the need to modify the core of the library. These rules are not expressible in terms of TypeScript typesystem and need to be enforced purely in the runtime.
2. Allow users to transform values before validation in order to support custom formats and inputs.
3. Support nominal types implemented by users.
4. Allow to bail out from further validation and assume value valid or invalid.

## Proposal

New higher order function `refine` provided by the library:

```ts
refine(refinement);
const refinement = (value, t) => {
  return value + 1; // continue validation on transformed value
  return t.left(value); // bail out with error
  return t.right(value); // bail out with success
  return t.right(value + 1); // bail out with success and transformed value
  return t.toEither(value > 10); // bail out with error or success depending on the condition
};
```

The result of calling `refine` is a function which can (optionally) take another schema as an argument and it might be passed to schemas:

```ts
const myRefinement = refine(refinement);
const validator = pipe(string, myRefinement, date, validate);

// alternatively
// validate(date(myRefinement(string())));
```

## Examples

```ts
const even = refine((value, t) => t.toEither(value % 2 === 0));

pipe(number, even, validate)(12); // 12
// validate(even(number()))(12); // 12

pipe(number, even, validate)(3); // ValidationError
// validate(even(number()))(3); // ValidationError
```

```ts
const noDuplicateItems = refine((arr, t) => {
  const allUnique = arr.every((item, index) => index === arr.indexOf(item));
  return allUnique ? arr : t.left(arr);
});

const uniqueStringArrayValidator = pipe(array(string), noDuplicateItems, validate);
// const uniqueStringArrayValidator = validate(noDuplicateItems(array(string())));
uniqueStringArrayValidator(['a', 'b']); // ['a', 'b']
uniqueStringArrayValidator(['a', 'b', 'a']); // ValidationError
```

```ts
const allowTimestamps = refine((value) => (Number.isInteger(value) ? new Date(value) : value));

pipe(allowTimestamps, date, validate)(1231231231231); // Date(2009-01-06T08:40:31.231Z)
// validate(date(allowTimestamps()))(1231231231231); // Date(2009-01-06T08:40:31.231Z)

pipe(allowTimestamps, date, validate)(Infinity); // ValidationError
// validate(date(allowTimestamps()))(Infinity); // ValidationError
```

```ts
const usDateString = refine((value) => {
  const result = typeof value === 'string' && value.match(/(?<m>\d\d)\/(?<d>\d\d)\/(?<y>\d\d\d\d)/);
  if (!result) {
    return value;
  }
  const { y, m, d } = result.groups;
  return `${y}-${m}-${d}`;
});

pipe(string, usDateString, date, validate)('01/12/2021'); // Date(2021-01-12T00:00:00.000Z)
// validate(date(usDateString(string())))('01/12/2021'); // Date(2021-01-12T00:00:00.000Z)

pipe(string, usDateString, date, validate)('14/12/2021'); // ValidationError
// validate(date(usDateString(string())))('14/12/2021'); // ValidationError
```

```ts
const integer = refine<Integer>((value, t) => t.toEither(Number.isInteger(value)));

pipe(number, integer, validate)(1); // Integer
// validate(integer(number()))(1); // Integer

pipe(number, integer, validate)(1.5); // ValidationError
// validate(integer(number()))(1.5); // ValidationError
```

```ts
const email = refine<Email>((value, t) => t.toEither(value.includes('@')));

pipe(string, email, validate)('hello@example.com'); // Email 'hello@example.com'
// validate(email(string()))('hello@example.com'); // Email 'hello@example.com'

pipe(string, email, validate)('just hello'); // ValidationError
// validate(email(string()))('just hello'); // ValidationError
```

```ts
const allowFalse = refine((value, t) => (value === false ? t.right(value) : value));

pipe(allowFalse, string, validate())(false); // false
// validate(string(allowFalse()))(false); // false

pipe(allowFalse, string, validate())(true); // ValidationError
// validate(string(allowFalse()))(true); // ValidationError
```

```ts
const isInObject = <T extends object>(object: T, key: keyof any): key is keyof T =>
  Object.prototype.hasOwnProperty.call(object, key);

const hasProperties = (properties: string[]) =>
  refine((value, t) =>
    properties.every((prop) => isInObject(value, prop)) ? value : t.left(value),
  );

pipe(hasProperties(['foo']), validate)({ foo: 'bar' }); // { foo: "bar" }
// validate(hasProperties(["foo"])())({ foo: "bar" }); // { foo: "bar" }

pipe(hasProperties(['foo']), validate)({ bar: 'foo' }); // ValidationError
// validate(hasProperties(["foo"])())({ bar: "foo" }); // ValidationError
```

```ts
const url = refine<URL>((value, t) => t.toEither(value.includes('https://')));

pipe(string, url, validate)('https://github.com/') // URL 'https://github.com/'
validate(url(string()))('https://github.com/') // URL 'https://github.com/'

pipe(string, url, validate)('abc') // ValidationError
validate(url(string()))('abc') // ValidationError
```

</details>
<hr>

## Previous versions

<details>
<summary>Version 1, 2020-01-26</summary>
### Goal 1

1. Introduce a mechanism which will allow users to extend schema with their own rules of validation without the need to modify the core of the library. These rules are not expressible in terms of TypeScript typesystem and need to be enforced purely in the runtime.

```ts
validate(evenNumber)(12); // 12
validate(evenNumber)(3); // ValidationError
```

```ts
validate(noDuplicateItems)(['a', 'b']); // ['a', 'b']
validate(noDuplicateItems)(['a', 'b', 'a']); // ValidationError
```

### Goal 2

2. Allow users to transform values before validation in order to support custom formats and inputs.

```ts
validate(timestamp(date()))(1231231231231); // Date(2009-01-06T08:40:31.231Z)
validate(timestamp(date()))(Infinity); // ValidationError
```

```ts
validate(weirdDate(date()))('01/12/2021'); // Date(2021-01-12T00:00:00.000Z)
validate(weirdDate(date()))('14/12/2021'); // ValidationError
```

### Goal 3

3. Support nominal types implemented by users.

```ts
type Integer = Nominal<number>; // `Nominal` is provided by user
const integer = (x: unknown): x is Integer => { ?? } // shape of this function is to be designed
validate(integer())(1); // Integer
validate(integer())(1.5); // ValidationError
```

```ts
type Email = Nominal<string>; // `Nominal` is provided by user
const email = (x: unknown): x is Email => { ?? } // shape of this function is to be designed
validate(email())('hello@example.com'); // Email
validate(email())('just hello'); // ValidationError
```

### Goal 4

4. Allow to bail out from further validation and assume value valid or invalid.

```ts
validate(allowFalse(string()))(false); // false
validate(allowFalse(string()))(true); // ValidationError
```

## Suggestion

New higher order function `refine` and `refineAfter` provided by the library:

```ts
refineAfter(beforeSchema)(refinement);
refine(refinement);
const refinement = (value, t) => {
  return value + 1; // continue validation on transformed value
  return t.left(value); // bail out with error
  return t.right(value); // bail out with success
  return t.right(value + 1); // bail out with success and transformed value
  return t.toEither(value > 10); // bail out with error or success depending on the condition
};
```

## Examples

```ts
const evenNumber = refineAfter(number())((value, t) => t.toEither(value % 2 === 0));
validate(evenNumber)(12); // 12
validate(evenNumber)(3); // ValidationError
```

```ts
const noDuplicateItems = (arr, t) => {
  const allUnique = arr.every((item, index) => index === arr.indexOf(item));
  return allUnique ? arr : t.left(arr);
};
const uniqueStringArray = refineAfter(array(string()))(noDuplicateItems);
validate(uniqueStringArray)(['a', 'b']); // ['a', 'b']
validate(uniqueStringArray)(['a', 'b', 'a']); // ValidationError
```

```ts
const withTimestamps = refine((value) => (Number.isInteger(value) ? new Date(value) : value));
validate(withTimestamps(date()))(1231231231231); // Date(2009-01-06T08:40:31.231Z)
validate(withTimestamps(date()))(Infinity); // ValidationError
```

```ts
const parseUsDateString = (value: string) => {
  const result = value.match(/(?<m>\d\d)\/(?<d>\d\d)\/(?<y>\d\d\d\d)/);
  if (!result) {
    return null;
  }
  const { y, m, d } = result.groups;
  return `${y}-${m}-${d}`;
};
const weirdDate = refineAfter(string())((value) => parseUsDateString(value) ?? value);
validate(weirdDate(date()))('01/12/2021'); // Date(2021-01-12T00:00:00.000Z)
validate(weirdDate(date()))('14/12/2021'); // ValidationError
```

```ts
const integer = refineAfter<Integer>(number())((value, t) => t.toEither(Number.isInteger(value)));
validate(integer())(1); // Integer
validate(integer())(1.5); // ValidationError
```

```ts
const email = refineAfter<Email>(string())((value, t) => t.toEither(value.includes('@')));
validate(email())('hello@example.com'); // Email
validate(email())('just hello'); // ValidationError
```

```ts
const allowFalse = refine((value, t) => (value === false ? t.right(value) : value));
validate(allowFalse(string()))(false); // false
validate(allowFalse(string()))(true); // ValidationError
```

</details>

**Comment by iamkacperwasik** [comment]
```ts
type Key = string | number | symbol;

const isInObject = (object: Object, key: Key) => {
  return object.hasOwnProperty(key);
};

const hasProperties = refineAfter(array(object({}), array(string(), number())))(
  ([object, keys]: [object: {}, keys: Key[]], t) => {
    return t.toEither(keys.every((key) => isInObject(object, key)));
  }
);

validate(hasProperties)([{ foo: "bar" }, ["foo"]]); 
validate(hasProperties)([{ bar: "foo" }, ["foo"]]); // <- error
```

**Comment by mmiszy** [comment]
With current proposal it would be:

```ts
const isInObject = <T extends object>(
  object: T,
  key: keyof any
): key is keyof T => Object.prototype.hasOwnProperty.call(object, key);

const hasProperties = (properties: string[]) =>
  refineAfter((value, t) =>
    properties.every((prop) => isInObject(value, prop)) ? value : t.left(value)
  )();

validate(hasProperties(["foo"]))({ foo: "bar" }); // { foo: "bar" }
validate(hasProperties(["foo"]))({ bar: "foo" }); // ValidationError
```

**Comment by wisnie** [comment]
```ts
type URL = Nominal<string>
const url = refineAfter<Email>(string())((value, t) => t.toEither(value.includes('https://')));
validate(url())('https://github.com/') // Returns 'https://github.com/'
validate(url())('abc') // Throws ValidationError
```

**Comment by mmiszy** [comment]
I've updated the RFC with a new simplified proposal:

- focus on the `pipe` helper for readability
- remove `refineAfter` as it was confusing
- allow schemas to take a refinement as an argument to post-validate
- allow refinements to take a schema as an argument to pre-validate

https://github.com/typeofweb/schema/issues/45#issue-794322642

**Comment by o-alexandrov** [comment]
Please kindly share the state of the [refinements](https://github.com/typeofweb/schema/tree/refinements) branch.
I understand you might change the API, but is it in a working state to attempt to test it?

**Comment by mmiszy** [comment]
@o-alexandrov not yet! I'll share an update when it is.

**Comment by mmiszy** [comment]
@o-alexandrov I've just published a first `nightly` release (0.6.1-0) to npmjs. It's working, it's fast, the docs are not updated yet and there are some breaking changes. Thankfully, TypeScript helps a lot so you can test it if you want :) There are lots of examples in https://github.com/typeofweb/schema/blob/ae724b91d6adab2d4bc993fdf28841d3c65db6bb/__tests__/refinements.spec.ts#L15

**Comment by o-alexandrov** [comment]
Thank you very much. I’ll try it out for one of the clients, [ERinfo](https://erinfo.me/), and report findings here.

I truly hope this project gets more of the community support as imho, you came up with the most lightweight (by an extreme margin) validation with static type inference

**Comment by o-alexandrov** [comment]
Tried `@typeofweb/schema` package in one of the projects and it was a great experience. Thank you.
Refinements are great. Everything was natural and easy to understand based on the test scenarios [you mentioned above](https://github.com/typeofweb/schema/issues/45#issuecomment-786789753).

The only minor aspects that I had to overcome are:
- the docs currently don't specify the need to invoke an object and some other random schemas, but then I looked at your PR's changes and understood the need to define a schema differently:
  - `object({ ... })()` instead of how it is in the docs `object({ ... })`
- I don't see a built-in way to throw custom `error` messages with Refinements, so I had to create a helper

I also stumbled upon two minor nuances with `validator`, but I will create a separate issue discussing it and how I currently patched on my end
